File: 2025-9322-team04_finproject_java/src/main/java/Client_Java/player/model/GameProperModel.java
package Client_Java.player.model;

import Client_Java.player.PlayerCorbaManager;

public class GameProperModel {
    private String wordToGuess;
    private boolean[] revealedLetters;
    private int lives;
    private int timerSeconds;
    private int remainingGuesses;

    private final PlayerCorbaManager corbaManager;

    public GameProperModel(String word) {
        this.corbaManager = PlayerCorbaManager.getInstance();
        this.wordToGuess = word.toUpperCase();
        this.revealedLetters = new boolean[word.length()];
        this.lives = 5;  // Default lives
        this.timerSeconds = corbaManager.getAdminService().getRoundDuration();
    }

    public String getWordToGuess() {
        return wordToGuess;
    }


    public boolean[] getRevealedLetters() {
        return revealedLetters;
    }


    public boolean isLetterInWord(char letter) {
        boolean found = false;
        for (int i = 0; i < wordToGuess.length(); i++) {
            if (wordToGuess.charAt(i) == letter) {
                revealedLetters[i] = true;  // Mark letter as revealed
                found = true;
            }
        }
        return found;
    }


    public int getLives() {
        return lives;
    }


    public void decrementLives() {
        if (lives > 0) lives--;
    }


    public int getTimerSeconds() {
        return timerSeconds;
    }


    public void decrementTimer() {
        timerSeconds--;
    }


    public boolean isGameOver() {
        return lives <= 0 || isWordFullyRevealed();
    }


    public boolean isWordFullyRevealed() {
        for (boolean b : revealedLetters) {
            if (!b) return false;
        }
        return true;
    }

    public void setWordToGuess(String word) {
        this.wordToGuess = word.toUpperCase();
        this.revealedLetters = new boolean[word.length()];
    }

    public void resetGame() {
        this.lives = 5;
        this.timerSeconds = 30;
        this.revealedLetters = new boolean[wordToGuess.length()];
    }

    public void updateWordState(String maskedWord, String actualWord, short remainingGuesses) {
        this.wordToGuess = actualWord;
        this.remainingGuesses = remainingGuesses;
        this.revealedLetters = new boolean[actualWord.length()];
        for (int i = 0; i < maskedWord.length(); i++) {
            if (maskedWord.charAt(i) != '_') {
                revealedLetters[i] = true;
            }
        }
    }

    public boolean isSessionValid() {
        String sessionId = PlayerCorbaManager.getInstance().getSessionId();
        return PlayerCorbaManager.getInstance().getLoginService().isSessionActive(sessionId);
    }
}

File: 2025-9322-team04_finproject_python/Client_Python/player/model/GameProperModel.py
from Client_Python.manager.PlayerCorbaManager import PlayerCorbaManager

class GameProperModel:
    def __init__(self, word: str):
        orb_args = ["-ORBInitRef", "NameService=corbaloc::localhost:4321/NameService"]
        self.manager = PlayerCorbaManager.get_instance(orb_args)
        #print(f"Model initialized with word: {word}") #TO CHECK WHAT WORD IS INITIALIZED
        self.word_to_guess = word.upper()
        self.revealed_letters = [False] * len(word)
        self.lives = 5
        self.timer_seconds = PlayerCorbaManager.get_instance().get_admin_service().getRoundDuration()

    def is_letter_in_word(self, letter: str) -> bool:
        return letter in self.word_to_guess

    def is_word_fully_revealed(self) -> bool:
        return all(self.revealed_letters)

    def decrement_lives(self):
        if self.lives > 0:
            self.lives -= 1

    def decrement_timer(self):
        if self.timer_seconds > 0:
            self.timer_seconds -= 1

    def get_word_to_guess(self) -> str:
        return self.word_to_guess

    def get_revealed_letters(self):
        return self.revealed_letters

    def get_lives(self) -> int:
        return self.lives

    def get_timer_seconds(self) -> int:
        return self.timer_seconds
